pkg/generate/cspc.go

```
@@ -0,0 +1,214 @@
+	"github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// isPoolValid checks if a CStor pool is valid
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Pool calls the generate routine for different cas-types
+func Pool(nodes []string, devs int, raid string) error {
+	c := client.NewK8sClient()
+	if !isPoolTypeValid(strings.ToLower(raid)) {
+		// TODO: Use the well defined pool constant types from openebs/api when added there
+		return fmt.Errorf("invalid pool type %s", raid)
+	}
+	_, str, err := CSPC(c, nodes, devs, strings.ToLower(raid))
+	if err != nil {
+		return err
+	}
+	fmt.Println(str)
+	return nil
+}
+
+// CSPC takes eligible nodes, number of devices and poolType to create a pool cluster template
+func CSPC(c *client.K8sClient, nodes []string, devs int, poolType string) (*cstorv1.CStorPoolCluster, string, error) {
+	// 0. Figure out the OPENEBS_NAMESPACE for CStor
+	cstorNS, err := c.GetOpenEBSNamespace(util.CstorCasType)
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
+	}
+	// 1. Validate nodes & poolType, fetch disks
+	nodeList, err := c.GetNodes(nodes, "", "")
+	if err != nil {
+		return nil, "", fmt.Errorf("(server error) unable to fetch nodes %s", err)
+	}
+	if len(nodeList.Items) != len(nodes) {
+		return nil, "", fmt.Errorf("not all worker nodes are available for provisioning a CSPC")
+	}
+	// 2. Fetch BD's from the eligible/valid nodes
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(nodes, ",")+")")
+	if err != nil || len(bds.Items) == 0 {
+		return nil, "", fmt.Errorf("no blockdevices found in %s nodes", nodes)
```

Let's also log about the error and do we need to change the format specifier to %v since nodes is a string slice

---

this error will get logged eventually as this line of errors aren't swallowed by the line of callers